### PR TITLE
Speed up data extractor

### PR DESCRIPTION
**Pull request by alan-stokes**
gives us a 5.4 times speed up on simple synfire chain. likely to give more for larger sims. 

depends upon the following prs:

- [x] https://github.com/SpiNNakerManchester/SpiNNFrontEndCommon/pull/229
- [x] https://github.com/SpiNNakerManchester/PACMAN/pull/131
- [ ] https://github.com/SpiNNakerManchester/sPyNNaker7/pull/41
- [ ] https://github.com/SpiNNakerManchester/sPyNNaker8/pull/84
- [x] https://github.com/SpiNNakerManchester/SpiNNMan/pull/100
- [x] https://github.com/SpiNNakerManchester/SpiNNMachine/pull/53
- [ ] https://github.com/SpiNNakerManchester/SpiNNakerGraphFrontEnd/pull/71

Note that this work is not finished, but is stable. Future work will come which will increase this even further

### CODE REVIEW

File: spynnaker/pyNN/utilities/extracted_data.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    def get(self, projection, attribute):
+        """ allows getting data from a given projection and attribute
+
+        :param projection: the projection data was extracted from
+        :param attribute: the attribute to retrieve
+        :return: the attribute data in a connection holder
+        """
+        if projection in self._data:
+            if attribute in self._data[projection]:
+                return self._data[projection][attribute]
```
Comment (dkfellows) [inline spynnaker/pyNN/utilities/extracted_data.py]: Please add an explicit default return for when the conditions are not true.

File: spynnaker/pyNN/models/neuron/synaptic_manager.py [hunk 2]
```diff
@@ -869,26 +882,45 @@ def _retrieve_synaptic_block(
...
+                    # read in the synaptic row data
+                    single_block = numpy.asarray(data_receiver.get_data(
+                        transceiver, sender_extra_monitor_core_placement,
+                        direct_synapses_address + synaptic_block_offset,
+                        synaptic_block_size), dtype="uint8").view("uint32")
+                else:
+                    single_block = numpy.asarray(transceiver.read_memory(
+                        placement.x, placement.y,
+                        direct_synapses_address + synaptic_block_offset,
+                        synaptic_block_size), dtype="uint8").view("uint32")
```
Comment (dkfellows) [inline spynnaker/pyNN/models/neuron/synaptic_manager.py]: I see this pattern in multiple places in this method. Refactoring opportunity?

File: spynnaker/pyNN/abstract_spinnaker_common.py [hunk 3]
```diff
@@ -347,6 +349,96 @@ def reset_number_of_neurons_per_core(self):
...
+            for edge in edges:
+                placement = self._placements.get_placement_of_vertex(
+                    edge.post_vertex)
+                chip = self._machine.get_chip_at(placement.x, placement.y)
+
+                # locate extra monitor cores on the board of this chip
+                extra_monitor_cores_on_board = set()
+                for (chip_x, chip_y) in self._machine.get_chips_on_board(chip):
+                    extra_monitor_cores_on_board.add(
+                        extra_monitors_per_chip[(chip_x, chip_y)])
```
Comment (dkfellows) [inline spynnaker/pyNN/abstract_spinnaker_common.py]: It seems wasteful to untuple the x,y just to retuple them up again just below.

Comment (alan-stokes) [reply spynnaker/pyNN/abstract_spinnaker_common.py]: fixed

File: spynnaker/pyNN/models/neuron/synaptic_manager.py (hunk 2 above)
Comment (alan-stokes) [reply spynnaker/pyNN/models/neuron/synaptic_manager.py]: annoyingly you gain nothing doing the refactor. its actually more lines. as your refactoring one line in the first place. most of it is parameters.

File: spynnaker/pyNN/utilities/extracted_data.py (hunk 1 above)
Comment (alan-stokes) [reply spynnaker/pyNN/utilities/extracted_data.py]: done

**Pull request merged**